### DISCUSSION TITLE
Not able to send password recovery mails to the users

### DISCUSSION THREAD

**Discussion by himanshuZee** [ory/kratos#2822] [Q&A]
Users not receiving emails even though i get a success response for self-service/recovery?flow={flowId}
I have tested with following courier configurations using sendgrid :
1. connection_uri: smtps://apikey:**********@smtp.sendgrid.net:25/?skip_ssl_verify=false
2. connection_uri: smtp://apikey:**********@smtp.sendgrid.net:25/
3. connection_uri: smtp://apikey:**********@smtp.sendgrid.net:25/?disable_starttls=true
4. connection_uri: smtps://apikey:**********@smtp.sendgrid.net:25/
5. connection_uri: smtps://apikey:**********@smtp.sendgrid.net:25/?skip_ssl_verify=true

I'm trying to test the functionality on localhost first before deploying to our dev and staging environments.
here is the log exactly when it says sending out email
{
level=info msg=Sending out recovery email with recovery link. audience=audit email_address=himanshu9821@gmail.com identity_id=7e8dc0fc-c4f2-455b-ad0c-321894a50b8e recovery_link_id=eeeeacee-e69f-4799-a2f2-7f7aa8239943 recovery_link_token=rSEzmlrW2kJ6TQ1HY12XOvXyUoyzepi8 service_name=Ory Kratos service_version=master via=email
}

Please help with the right config for courier.

**Comment by aeneasr** [comment]
Have you checked https://www.ory.sh/docs/kratos/emails-sms/sending-emails-smtp#send-emails-using-your-smtp-server ?

**Reply by mlb5000** [reply]
I've having the same issue (described below). Do you have any insight?

**Comment by mlb5000** [comment]
I'm having the exact same issue. Kratos logs report no errors in sending the email. Giving it a deliberately wrong username, password, or SMTP host it doesn't raise any errors either. But if you omit `connection_uri` from the `courier` configuration section, the service will not start (as expected). So it's reading the configuration value at startup, but doesn't appear to ever actually try sending the emails.

For me this is on Kratos v0.13.0.

**Reply by vinckr** [reply]
Hello @mlb5000 
Can you try upgrading to the latest version? We don't really support the older versions, so maybe its related to a bug or was something else that got fixed in 1.0

**Comment by mlb5000** [comment]
@vinckr same issue in 1.0.0. No emails are sent. I've tried the same SMTP credential variations as the original poster, have tried giving it a bad SMTP hostname, bad username, bad password. None of them raise any errors, but it claims every single time `Sending out recovery email with recovery code.`

```
kratos-9dc587654-74l4n kratos time=2023-10-19T13:24:53Z level=debug msg=Preparing recovery code. address=matt@******.com audience=application service_name=Ory Kratos service_version=v1.0.0 via=email
kratos-9dc587654-74l4n kratos time=2023-10-19T13:24:53Z level=info msg=Sending out recovery email with recovery code. audience=audit email_address=matt@******.com identity_id=887fee45-4cd5-******-******-****** recovery_code=****** recovery_code_id=b7e07e60-764b-******-******-****** service_name=Ory Kratos service_version=v1.0.0 via=email
```

**Reply by jonas-jonas** [reply]
Are you starting Kratos with the --watch-courier flag or the courier as its own pod/process? See here for more info: https://www.ory.sh/docs/kratos/self-hosted/mail-courier-selfhosted

**Reply by mlb5000** [reply]
@jonas-jonas well now we're making progress at least. Not sure why it printed it twice. Still no errors or email though (also checked the spam folder).

```
kratos-5d476f4c65-htb5q kratos time=2023-10-19T13:38:15Z level=debug msg=Courier sent out message. audience=application message_id=5ae062ae-54e1-4891-b7af-4396a5188798 message_nid=f8dc319a-ba33-4e71-b9a6-effb20fe82ea message_subject=Recover access to your account message_template_type=recovery_code_valid message_type=email service_name=Ory Kratos service_version=v1.0.0
kratos-5d476f4c65-htb5q kratos time=2023-10-19T13:38:15Z level=debug msg=Courier sent out message. audience=application message_id=5ae062ae-54e1-4891-b7af-4396a5188798 message_nid=f8dc319a-ba33-4e71-b9a6-effb20fe82ea message_subject=Recover access to your account message_template_type=recovery_code_valid message_type=email service_name=Ory Kratos service_version=v1.0.0
```

**Reply by mlb5000** [reply]
@jonas-jonas I spoke too soon. Just took an excessive amount of time to arrive. It _did_ send me two recovery codes though.

I've tried it a few times now, and it does seem to be sending me two recovery emails each time I submit the form once, each with a different code.

**Reply by jonas-jonas** [reply]
> Just took an excessive amount of time to arrive.

Likely due to spam filters, etc.

>I've tried it a few times now, and it does seem to be sending me two recovery emails each time I submit the form once, each with a different code.

That is odd, are you starting the Kratos with the `--watch-courier` flag or as a standalone process?

**Reply by mlb5000** [reply]
`--watch-courier`. Is it possible that previous recovery requests that were not handled (because no courier service was listening) got backed up and pushed through only when the next request came through? I tried it a few more times and then it started only sending the one email each time. If I had to guess, it was sending "old" recovery code emails that just got backed up.

Now my issue is with Account Verification not sending. Should I make another post for that one, or drop the details here?

**Reply by jonas-jonas** [reply]
Yes, could be. But the log messaged indicated that it was sending the same message twice, which would obviously also be odd.

No, do share here. It's the same code path, so there shouldn't be any differences.

**Reply by mlb5000** [reply]
@jonas-jonas Well, this one might actually be an internal issue

```
kratos-5d476f4c65-htb5q kratos time=2023-10-19T13:51:25Z level=debug msg=Preparing verification code. address=matt@******.com audience=application service_name=Ory Kratos service_version=v1.0.0 via=email
kratos-5d476f4c65-htb5q kratos time=2023-10-19T13:51:25Z level=info msg=Address verification was requested for an unknown address. audience=audit email_address=<nil> service_name=Ory Kratos service_version=v1.0.0 strategy=code via=email was_notified=false
```

Notice how the second log message says `email_address=<nil>`, despite the first one having it correct? That seems odd. I verified that the `identity` and `email` are set on my account, and that they match, but it still claims that it's an "unknown address".

**Reply by jonas-jonas** [reply]
Well, that's actually just a bug in the log message. Nice find!

Does the email address you're trying to verify belong to a registered account?

**Reply by mlb5000** [reply]
@jonas-jonas yes, it does. I verified that both the `identity` and `email` fields on my account match the email I'm entering there, but it still claims it doesn't exist.

**Reply by jonas-jonas** [reply]
Could you post your identity schema?

**Comment by joshdoesthis** [comment]
I was having trouble with this as well until I read the following: https://www.ory.sh/docs/kratos/cli/kratos-courier-watch
E.g. serve -c /etc/config/kratos/kratos.yml --watch-courier
